pseudopotential_calculator/calculations/adsorbate.py

```diff
@@ -0,0 +1,121 @@
+from __future__ import annotations
+
+from typing import TYPE_CHECKING, Literal, Self, cast
+
+import matplotlib.pyplot as plt
+import numpy as np
+from ase import Atom, Atoms
+from matplotlib import cm
+
+from pseudopotential_calculator.calculations.generic import (
+    OptimizationParamsBase,
+    repeat_cell,
+)
+from pseudopotential_calculator.calculations.slab import get_n_th_slab_layer_height
+from pseudopotential_calculator.castep import (
+    CastepConfig,
+    get_calculator_atom,
+    get_default_calculator,
+)
+
+if TYPE_CHECKING:
+    from ase.calculators.castep import Castep
+    from matplotlib.figure import Figure
+
+
+class AdsorbateOptimizationParams(OptimizationParamsBase):
+    """Parameters of a adsorbate calculation."""
+
+    @property
+    def kpoint_mp_grid(self: Self) -> str:
+        return f"{self.n_k_points} {self.n_k_points} {1}"
+
+
+def get_adsorbate_optimization_calculator(
+    atoms: Atoms,
+    parameters: AdsorbateOptimizationParams,
+    config: CastepConfig,
+    task: Literal["Energy", "GeometryOptimization"] = "Energy",
+) -> Castep:
+    calculator = get_default_calculator(config)
+
+    calculator.param.xc_functional = parameters.xc_functional
+    calculator.param.cut_off_energy = parameters.cut_off_energy
+    calculator.param.spin_polarized = parameters.spin_polarized
+    calculator.cell.kpoint_mp_grid = parameters.kpoint_mp_grid
+
```

Comment:
also add symmetry_generate

Reply:
what about calculator.param.max_scf_cycles = parameters.max_scf_cycles? do we add this and calculator.cell.symmetry_generate = True everywhere since you said it's to improve castep calculation

Reply:
yeah add it everywhere - symmetry_generate is just an optimization, and max_scf_cycles wont change the calculation, but we will sometimes need to adjust it to ensure we converge